File: sdk/samples/all in on code/Visual Studio 2008/CSSL3DeepZoom/CSSL3DeepZoom/MouseWheelHelper.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.Windows.Browser;

namespace CSSL3DeepZoom
{
        // Courtesy of Pete Blois
        public class MouseWheelEventArgs : EventArgs
        {
            private double delta;
            private bool handled = false;

            public MouseWheelEventArgs(double delta)
            {
                this.delta = delta;
            }

            public double Delta
            {
                get { return this.delta; }
            }

            // Use handled to prevent the default browser behavior!
            public bool Handled
            {
                get { return this.handled; }
                set { this.handled = value; }
            }
        }

        public class MouseWheelHelper
        {

            public event EventHandler<MouseWheelEventArgs> Moved;
            private static Worker worker;
            private bool isMouseOver = false;

            public MouseWheelHelper(FrameworkElement element)
            {

                if (MouseWheelHelper.worker == null)
                    MouseWheelHelper.worker = new Worker();

                MouseWheelHelper.worker.Moved += this.HandleMouseWheel;

                element.MouseEnter += this.HandleMouseEnter;
                element.MouseLeave += this.HandleMouseLeave;
                element.MouseMove += this.HandleMouseMove;
            }

            private void HandleMouseWheel(object sender, MouseWheelEventArgs args)
            {
                if (this.isMouseOver)
                    this.Moved(this, args);
            }

            private void HandleMouseEnter(object sender, EventArgs e)
            {
                this.isMouseOver = true;
            }

            private void HandleMouseLeave(object sender, EventArgs e)
            {
                this.isMouseOver = false;
            }

            private void HandleMouseMove(object sender, EventArgs e)
            {
                this.isMouseOver = true;
            }

            private class Worker
            {

                public event EventHandler<MouseWheelEventArgs> Moved;

                public Worker()
                {

                    if (HtmlPage.IsEnabled)
                    {
                        HtmlPage.Window.AttachEvent("DOMMouseScroll", this.HandleMouseWheel);
                        HtmlPage.Window.AttachEvent("onmousewheel", this.HandleMouseWheel);
                        HtmlPage.Document.AttachEvent("onmousewheel", this.HandleMouseWheel);
                    }

                }

                private void HandleMouseWheel(object sender, HtmlEventArgs args)
                {
                    double delta = 0;

                    ScriptObject eventObj = args.EventObject;

                    if (eventObj.GetProperty("wheelDelta") != null)
                    {
                        delta = ((double)eventObj.GetProperty("wheelDelta")) / 120;


                        if (HtmlPage.Window.GetProperty("opera") != null)
                            delta = -delta;
                    }
                    else if (eventObj.GetProperty("detail") != null)
                    {
                        delta = -((double)eventObj.GetProperty("detail")) / 3;

                        if (HtmlPage.BrowserInformation.UserAgent.IndexOf("Macintosh") != -1)
                            delta = delta * 3;
                    }

                    if (delta != 0 && this.Moved != null)
                    {
                        MouseWheelEventArgs wheelArgs = new MouseWheelEventArgs(delta);
                        this.Moved(this, wheelArgs);

                        if (wheelArgs.Handled)
                            args.PreventDefault();
                    }
                }
            }
        }
}

File: sdk/samples/all in on code/Visual Studio 2008/VBSL3DeepZoom/VBSL3DeepZoom/MouseWheelHelper.vb
Imports System.Windows.Browser

Public Class MouseWheelHelper
	' Events
	Public Event Moved As EventHandler(Of MouseWheelEventArgs)

	' Methods
	Public Sub New(ByVal element As FrameworkElement)
		If (MouseWheelHelper.workerInstance Is Nothing) Then
			MouseWheelHelper.workerInstance = New Worker
		End If
		AddHandler MouseWheelHelper.workerInstance.Moved, New EventHandler(Of MouseWheelEventArgs)(AddressOf Me.HandleMouseWheel)
		AddHandler element.MouseEnter, New MouseEventHandler(AddressOf Me.HandleMouseEnter)
		AddHandler element.MouseLeave, New MouseEventHandler(AddressOf Me.HandleMouseLeave)
		AddHandler element.MouseMove, New MouseEventHandler(AddressOf Me.HandleMouseMove)
	End Sub

	Private Sub HandleMouseEnter(ByVal sender As Object, ByVal e As EventArgs)
		Me.isMouseOver = True
	End Sub

	Private Sub HandleMouseLeave(ByVal sender As Object, ByVal e As EventArgs)
		Me.isMouseOver = False
	End Sub

	Private Sub HandleMouseMove(ByVal sender As Object, ByVal e As EventArgs)
		Me.isMouseOver = True
	End Sub

	Private Sub HandleMouseWheel(ByVal sender As Object, ByVal args As MouseWheelEventArgs)
		If Me.isMouseOver Then
			RaiseEvent Moved(Me, args)
		End If
	End Sub


	' Fields
	Private isMouseOver As Boolean = False
	Private Shared workerInstance As Worker

	' Nested Types
	Private Class Worker
		' Events
		Public Event Moved As EventHandler(Of MouseWheelEventArgs)

		' Methods
		Public Sub New()
			If HtmlPage.IsEnabled Then
				HtmlPage.Window.AttachEvent("DOMMouseScroll", New EventHandler(Of HtmlEventArgs)(AddressOf Me.HandleMouseWheel))
				HtmlPage.Window.AttachEvent("onmousewheel", New EventHandler(Of HtmlEventArgs)(AddressOf Me.HandleMouseWheel))
				HtmlPage.Document.AttachEvent("onmousewheel", New EventHandler(Of HtmlEventArgs)(AddressOf Me.HandleMouseWheel))
			End If
		End Sub

		Private Sub HandleMouseWheel(ByVal sender As Object, ByVal args As HtmlEventArgs)
			Dim delta As Double = 0
			Dim eventObject As ScriptObject = args.EventObject
			If (Not eventObject.GetProperty("wheelDelta") Is Nothing) Then
				delta = (CDbl(eventObject.GetProperty("wheelDelta")) / 120)
				If (Not HtmlPage.Window.GetProperty("opera") Is Nothing) Then
					delta = -delta
				End If
			ElseIf (Not eventObject.GetProperty("detail") Is Nothing) Then
				delta = (-CDbl(eventObject.GetProperty("detail")) / 3)
				If (HtmlPage.BrowserInformation.UserAgent.IndexOf("Macintosh") <> -1) Then
					delta = (delta * 3)
				End If
			End If
			If (delta <> 0) Then
				Dim e As New MouseWheelEventArgs(delta)
				RaiseEvent Moved(Me, e)
				If e.Handled Then
					args.PreventDefault()
				End If
			End If
		End Sub

	End Class
End Class

Public Class MouseWheelEventArgs
	Inherits EventArgs
	' Methods
	Public Sub New(ByVal delta As Double)
		Me.deltaField = delta
	End Sub


	' Properties
	Public ReadOnly Property Delta() As Double
		Get
			Return Me.deltaField
		End Get
	End Property

	Public Property Handled() As Boolean
		Get
			Return Me.handledField
		End Get
		Set(ByVal value As Boolean)
			Me.handledField = value
		End Set
	End Property


	' Fields
	Private deltaField As Double
	Private handledField As Boolean = False
End Class

